src/Microsoft.Extensions.Logging.AzureWebApps/Internal/BackgroundSink.cs

```diff
@@ -0,0 +1,114 @@
+    {
+        /// <summary>
+        /// The default queue size
+        /// </summary>
+        public const int DefaultQueueSize = 1024;
+
+        private readonly CancellationTokenSource _disposedTokenSource = new CancellationTokenSource();
+        private readonly CancellationToken _disposedToken;
+
+        private readonly BlockingCollection<LogEvent> _messages;
+        private readonly Thread _workerThread;
+
+        private ILogEventSink _innerSink;
+
+        /// <summary>
+        /// Creates a new instance of the <see cref="BackgroundSink"/> class
+        /// </summary>
+        /// <param name="innerSink">The inner sink which does the actual logging</param>
+        /// <param name="maxQueueSize">The maximum size of the background queue</param>
+        public BackgroundSink(ILogEventSink innerSink, int maxQueueSize)
+        {
+            if (innerSink == null)
+            {
+                throw new ArgumentNullException(nameof(innerSink));
+            }
+
+            _disposedToken = _disposedTokenSource.Token;
+
+            if (maxQueueSize <= 0)
+            {
+                _messages = new BlockingCollection<LogEvent>(new ConcurrentQueue<LogEvent>());
+            }
+            else
+            {
+                _messages = new BlockingCollection<LogEvent>(new ConcurrentQueue<LogEvent>(), maxQueueSize);
+            }
+
+            _innerSink = innerSink;
+
+            _workerThread = new Thread(Worker);
+            _workerThread.Name = GetType().Name;
+            _workerThread.IsBackground = true;
+            _workerThread.Start();
+        }
+
+        /// <inheritdoc />
+        public void Emit(LogEvent logEvent)
+        {
+            if (!_disposedToken.IsCancellationRequested)
+            {
+                _messages.Add(logEvent);
+            }
+        }
+
+        /// <summary>
+        /// Disposes this object instance
+        /// </summary>
+        public virtual void Dispose()
+        {
+            if (!_disposedTokenSource.IsCancellationRequested)
+            {
+                lock (_disposedTokenSource)
+                {
+                    if (!_disposedTokenSource.IsCancellationRequested)
+                    {
+                        _disposedTokenSource.Cancel();
+                    }
+
+                    // Wait for the thread to complete before disposing the resources
+                    _workerThread.Join();
+                    _messages.Dispose();
+                }
+            }
+        }
+
+        private void Worker()
+        {
+            while (!_disposedToken.IsCancellationRequested)
+            {
+                // This block until either an item is available
+                // or the token is cancelled 
+                LogEvent logEvent = _messages.Take(_disposedToken);
+
+                if (!_disposedToken.IsCancellationRequested &&
+                    logEvent != null)
+                {
+                    PassLogEventToInnerSink(logEvent);
```

Comment:
This sounds like we are flushing the queued log events one at a time? we ideally want to avoid frequent disk access...not sure if I am missing something...

I was thinking we should flush log events when: queue is full or the logger provider is being disposed etc.